src/test/Dai.t.sol

```diff
@@ -1,316 +1,606 @@
+        assertTrue(token.transfer(address(0xBEEF), 1e18));
+        assertEq(token.totalSupply(), 1e18);
+
+        assertEq(token.balanceOf(address(this)), 0);
+        assertEq(token.balanceOf(address(0xBEEF)), 1e18);
     }
-    function doMint(address guy, uint wad) public {
-        token.mint(guy, wad);
+
+    function testTransferBadAddress() public {
+        token.mint(address(this), 1e18);
+
+        vm.expectRevert("Dai/invalid-address");
+        token.transfer(address(0), 1e18);
+        vm.expectRevert("Dai/invalid-address");
+        token.transfer(address(token), 1e18);
     }
-    function doBurn(address guy, uint wad) public {
-        token.burn(guy, wad);
+
+    function testTransferFrom() public {
+        address from = address(0xABCD);
+
+        token.mint(from, 1e18);
+
+        vm.prank(from);
+        token.approve(address(this), 1e18);
+
+        vm.expectEmit(true, true, true, true);
+        emit Transfer(from, address(0xBEEF), 1e18);
+        assertTrue(token.transferFrom(from, address(0xBEEF), 1e18));
+        assertEq(token.totalSupply(), 1e18);
+
+        assertEq(token.allowance(from, address(this)), 0);
+
+        assertEq(token.balanceOf(from), 0);
+        assertEq(token.balanceOf(address(0xBEEF)), 1e18);
     }
 
-}
+    function testTransferFromBadAddress() public {
+        token.mint(address(this), 1e18);
+        
+        vm.expectRevert("Dai/invalid-address");
+        token.transferFrom(address(this), address(0), 1e18);
+        vm.expectRevert("Dai/invalid-address");
+        token.transferFrom(address(this), address(token), 1e18);
+    }
 
-interface Hevm {
-    function warp(uint256) external;
-}
+    function testInfiniteApproveTransferFrom() public {
+        address from = address(0xABCD);
 
-contract DaiTest is DSTest {
-    uint constant initialBalanceThis = 1000;
-    uint constant initialBalanceCal = 100;
+        token.mint(from, 1e18);
 
-    Dai token;
-    Hevm hevm;
-    address user1;
-    address user2;
-    address self;
+        vm.prank(from);
+        vm.expectEmit(true, true, true, true);
+        emit Approval(from, address(this), type(uint256).max);
+        token.approve(address(this), type(uint256).max);
 
-    uint amount = 2;
-    uint fee = 1;
-    uint nonce = 0;
-    uint deadline = 0;
-    address cal = 0x29C76e6aD8f28BB1004902578Fb108c507Be341b;
-    address del = 0xdd2d5D3f7f1b35b7A0601D6A00DbB7D44Af58479;
-    bytes32 r = 0x3e8ffe622889f0560938c883bf1497b90d4c2fc73631cd5c50b9e7bfe1fbb2fc;
-    bytes32 s = 0x5b98e77b91f13c52fe540b945d11d9a96441133726082d4b6b1753341a3e52ff;
-    uint8 v = 28;
-    bytes32 _r = 0x85da10f8af2cf512620c07d800f8e17a2a4cd2e91bf0835a34bf470abc6b66e5;
-    bytes32 _s = 0x7e8e641e5e8bef932c3a55e7365e0201196fc6385d942c47d749bf76e73ee46f;
-    uint8 _v = 27;
+        vm.expectEmit(true, true, true, true);
+        emit Transfer(from, address(0xBEEF), 1e18);
+        assertTrue(token.transferFrom(from, address(0xBEEF), 1e18));
```

Comment:
Shouldn't we `stopPrank` before this? As otherwise we are executing the `transferFrom` from `from` so we wouldn't be testing that the `address(this)` allowance is not decreasing (if I'm understanding the intention of the test correctly).

Reply:
Ah nvm, `prank` is just for the *next* call.